### PR TITLE
IAM Entity protect

### PR DESCRIPTION
**Pull request by ayalganem**
* IAM entity has two mods of protection:
     - Protect/Restrict mod, implemented in AWS cloud account resource.
     - Protect with elevation mod, implemented in user resource.

### CODE REVIEW

File: services/cloudaccounts/aws/aws.go [hunk 1]
```diff
@@ -24,6 +26,11 @@ type AttachIamSafeRequest struct {
 	Data           Data   `json:"data"`
 }
 
+type RestrictedIamEntitiesRequest struct {
+	EntityName string `json:"entityName"` // aws iam user name or aws role role
```
Comment (erez-dome9) [inline services/cloudaccounts/aws/aws.go]: role name

File: services/cloudaccounts/aws/aws.go [hunk 2]
```diff
@@ -44,6 +51,21 @@ type CloudAccountResponse struct {
 	LambdaScanner          bool                    `json:"lambdaScanner"`
 }
 
+type ProtectIAMEntitiesResponse struct {
+	RolesArn []Arn `json:"rolesArns"`
+	UsersArn []Arn `json:"usersArns"`
+}
+
+type Arn struct {
```
Comment (erez-dome9) [inline services/cloudaccounts/aws/aws.go]: It's confusing to use this struct name `Arn` especially when you have a field called `Arn` inside. Why calling it Arn? Arn is an AWS string identifier. Lets not use it for different types

File: services/users/user.go [hunk 3]
```diff
@@ -1,16 +1,27 @@
...
 
 const (
 	userResourcePath      = "user"
 	userResourceOwnerPath = "account/owner"
+	userIAMSAfe           = "iam-safe"
+	userAccounts          = "accounts"
+	userIAMEntities       = "iamEntities"
 )
 
+var gUserEmailID = map[string]string{}
```
Comment (erez-dome9) [inline services/users/user.go]: what is `gUser` stands for?

File: services/users/user.go [hunk 4]
```diff
@@ -133,11 +166,179 @@ func (service *Service) SetUserAsOwner(userId string) (*http.Response, error) {
...
+	delete(gUserEmailID, user.Email)
 	return resp, err
 }
+
+/*
+	iam safe entities
+*/
+
+func (service *Service) ProtectWithElevationAWSIAMEntityCreate(d9CloudAccountID, entityName, entityType string, d9UsersID []string) (*ProtectIAMSafeEntitiesResponse, *http.Response, error) {
+	var srv = aws.New(service.Client.Config)
```
Comment (erez-dome9) [inline services/users/user.go]: `srv` -> `cloudAccountService` ? make `srv` meaningful

File: services/users/user.go [hunk 5]
```diff
@@ -133,11 +166,179 @@ func (service *Service) SetUserAsOwner(userId string) (*http.Response, error) {
...
+	}
+	for _, userID := range d9UsersID {
+		relativeURL := fmt.Sprintf("%s/%s/%s/%s/%s/%s", userResourcePath, userID, userIAMSAfe, userAccounts, d9CloudAccountID, userIAMEntities)
+		resp, err = service.Client.NewRequestDo("POST", relativeURL, nil, body, &v)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	return &v, resp, nil
```
Comment (erez-dome9) [inline services/users/user.go]: you are making several requests but return only the last `&v` and `resp`

File: services/users/user.go (hunk 3 above)
Comment (ayalganem) [reply services/users/user.go]: g stands for global

File: services/users/user.go [hunk 6]
```diff
@@ -224,11 +224,11 @@ func (service *Service) ProtectWithElevationAWSIAMEntityCreate(d9CloudAccountID,
 	return &v, resp, nil
```
Comment (erez-dome9) [inline services/users/user.go]: remove resp from return if not relevant

**Pull request merged**